### ISSUE TITLE
hook component can not been clicked in react-dev-tool if you want it works well.

### ISSUE THREAD

**Issue by fantasticsoul** [facebook/react#16519]
### issue type
bug

### issue description
look at the code blow (it is very very very simple).
```jsx
import React, { useState } from 'react';

export default function DemoCounter() {
  const [fnWrapper] = useState('fn');
  const [count, setCount] = useState(0);
  fnWrapper.__proto__.setCount = s => {
    setCount(s);
  }

 // assign fnWrapper.__proto__.setCount to callSetCount
  const callSetCount = fnWrapper.__proto__.setCount;

  return (
    <div style={{ border: '1px solid blue', margin: '8px' }}>
      count: {count}
      {/* this does not work if I open react-dev-tool and click the dom node */}
      <input value={count} onChange={e => fnWrapper.__proto__.setCount(e.currentTarget.value)} />

       {/* this always works not matter I open react-dev-tool and click the dom node or not */}
      <input value={count} onChange={e => callSetCount(e.currentTarget.value)} />
    </div>
  );
}

```

**Comment by threepointone** [comment]
This is not typical usage of hooks and I'm following it hard to follow. Could you use codesandbox to make a reproducing case please?

**Comment by fantasticsoul** [comment]
sorry for that I can not make a reproducing case in codesandbox, because in codesandbox the react-dev-tool can not show the react dom tree for its inner browser, it can just show codesandbox self's react dom tree......

but it is very easy to trace the bug, you can just create an app by create-react-app and then copy my code mentioned above to your app's file, just open the react-dev-tool and click the two buttons, you will know what I mean..

**Comment by fantasticsoul** [comment]
can any one help me to resolve this problem?

**Comment by bvaughn** [comment]
> sorry for that I can not make a reproducing case in codesandbox, because in codesandbox the react-dev-tool can not show the react dom tree for its inner browser, it can just show codesandbox self's react dom tree......

This is incorrect.

You can use the browser React DevTools extension in a Code Sandbox app if you open the app in its own (full screen) window. Code Sandbox also embeds a version of the React DevTools that you can use in the editor/iframe view.

**Comment by bvaughn** [comment]
That being said, this is extremely unusual example code, so it's not really high priority to investigate on our side. Why are you wrapping a hook updater function in a string prototype ?

Anyway I think the reason for this is because DevTools shallowly re-renders components that use hooks in order to build up the map of hook-to-value to show in the selected properties panel. When it does this re-rendering, it temporarily overrides the React's built-in hooks with [a version that collects this metadata](https://github.com/facebook/react/blob/master/packages/react-debug-tools/src/ReactDebugHooks.js). In your case, since you're mutating a prototype function every render, the more recent render (by DevTools) causes you to override it with the injected no-op value so that future calls don't actually update state.

You can verify this by only overriding the prototype function once:
```js
if (typeof fnWrapper.__proto__.setCount !== "function") {
  fnWrapper.__proto__.setCount = s => {
    setCount(s);
  };
}
```

But to be clear, this type of mutation is not supported or encouraged as React does not know about it and so will not schedule a re-render. Either use an inline function, use the state updater directly or use `useCallback` etc

**Comment by fantasticsoul** [comment]
I put a new [demo](https://codesandbox.io/s/test-react-dev-tool-bug-kpb1y) in sandbox
```js
import React, { useState, useCallback } from 'react';

const fnWrapper = { setCount: ()=>{}};

export default function DemoCounter() {
  console.log('render DemoCounter');
  const m = useState(0);

  // this is not ok
  // fnWrapper.setCount = s => {
  //   console.log('fnWrapper.setCount');
  //   m[1](s);
  // }

  // this is not ok also
  fnWrapper.setCount = useCallback((s)=>{
    m[1](s)
  }, [m]);

  return (
    <div style={{ border: '1px solid blue', margin: '8px' }}>
      count: {m[0]}
      {/* this does not work if I open react-dev-tool and click the dom node */}
      <input value={m[0]} onChange={e => fnWrapper.setCount(e.currentTarget.value)} />

       {/* this always works not matter I open react-dev-tool and click the dom node or not */}
      <input value={m[0]} onChange={e => m[1](e.currentTarget.value)} />
    </div>
  );
}
```
I'm so confused about this dev-tool bug....

**Comment by fantasticsoul** [comment]
I just assign the lastest setter to `fnWrapper.setCount` in every render period but I don't know why it does's work as I imagine.

**Comment by fantasticsoul** [comment]
the code like below also does't effect
```js
 fnWrapper.setCount = m[1]
```

**Comment by fantasticsoul** [comment]
In my demo I've writen the code you told but the alert never been triggered
```js
  if (typeof fnWrapper.setCount !== "function") {
    console.log(typeof fnWrapper.setCount);
    console.log(fnWrapper.setCount);
    alert('wow');
  }
```
